update @react-native-community/eslint-config

likern opened the pull request:
## Summary

I updated **dependencies** for `@react-native-community/eslint-config` and adjusted configs to align with updated versions.
### Main changes:

1. `babel-eslint` became [`@babel/eslint-parser`](https://github.com/babel/babel/tree/main/eslint/babel-eslint-parser)
2. a companion (recommended) [`@babel/eslint-plugin`](https://github.com/babel/babel/tree/main/eslint/babel-eslint-plugin) is used for `@babel/semi` rule (handles class properties)
3. `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` were updated to support new TypeScript syntax
4. `prettier` were also updated
5. `prettier/react` was removed according to [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier) documentation


## Changelog
## Test Plan

1. Create new react-native project
2. Add updated `@react-native-community/eslint-config` into it (`yarn add file://....`)
3. Run `yarn lint`
4. Check there is no errors or warnings

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

fabriziobertoglio1987 commented:
sorry, probably I'm doing something wrong.. What do you think? 

```bash
yarn run v1.22.5
$ eslint .

Oops! Something went wrong! :(

ESLint: 7.12.0

Error: Failed to load parser '@babel/eslint-parser' declared in '.eslintrc » ./packages/eslint-config-react-native-community/index.js#overrides[0]': Cannot find module '@babel/eslint-parser'
Require stack:
- /home/fabrizio/Documents/sourcecode/opensource/react-native/packages/eslint-config-react-native-community/index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:924:15)
    at Function.resolve (node:internal/modules/cjs/helpers:98:19)
    at Object.resolve (/home/fabrizio/Documents/sourcecode/opensource/react-native/node_modules/@eslint/eslintrc/lib/shared/relative-module-resolver.js:28:50)
    at ConfigArrayFactory._loadParser (/home/fabrizio/Documents/sourcecode/opensource/react-native/node_modules/@eslint/eslintrc/lib/config-array-factory.js:923:45)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/home/fabrizio/Documents/sourcecode/opensource/react-native/node_modules/@eslint/eslintrc/lib/config-array-factory.js:723:43)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/home/fabrizio/Documents/sourcecode/opensource/react-native/node_modules/@eslint/eslintrc/lib/config-array-factory.js:664:20)
    at _normalizeObjectConfigData.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/home/fabrizio/Documents/sourcecode/opensource/react-native/node_modules/@eslint/eslintrc/lib/config-array-factory.js:757:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
error Command failed with exit code 2.
```

likern commented:
@fabriziobertoglio1987 Hello 👋🏻 How did you tested this pull request?  I've provided more detailed step-by-step guide to be sure we typed the same commands.

### Create new React Native project
```bash
npx react-native init FreshReactNative
cd FreshReactNative/
yarn install # install dependencies
```

### Check @babel/eslint-parser doesn't yet exist
```bash
# should show that no file or directory exist
ls node_modules/@babel/eslint-parser
```

### Clone this repository and pull request
```bash
# in another directory
git clone git@github.com:facebook/react-native.git react-native-master
cd react-native-master/
yarn install
# checkout this pull request
gh pr checkout 31182
# git branch
# shows this branch update/eslint-config-react-native-community
```

### Go to eslint-config package
```bash
cd packages/eslint-config-react-native-community/
yarn install
# ls node_modules/@babel/eslint-parser/
# shows this package exist
```

### Pack this package to be used locally
```bash
yarn pack
# archive react-native-community-eslint-config-v3.0.0.tgz appeared
```

### Install this package to FreshReactNative
```bash
# cd to FreshReactNative project location
pwd
# /home/me/Projects/FreshReactNative in my case
less package.json
# it shows "@react-native-community/eslint-config": "^2.0.0" in `devDependencies` section
# your location to archive might be different
yarn add --dev ../../react-native-master/packages/eslint-config-react-native-community/react-native-community-eslint-config-v3.0.0.tgz
# it shows `success Saved 16 new dependencies`
# `@babel/eslint-parser@7.13.10` dependency was installed
```

### Check @babel/eslint-parser exist in FreshReactNative
```bash
# should show this directory exists
ls node_modules/@babel/eslint-parser
```

### Verify new eslint-config is used
```bash
less package.json
# it shows "@react-native-community/eslint-config": "../../react-native-master/packages/eslint-config-react-native-community/react-native-community-eslint-config-v3.0.0.tgz"
# your location to archive might be different
```

### Run eslint to check it's working
```bash
yarn lint
```

### Check eslint detects bad code
```js
# add this line into App.js
const test = bool == '';
```
```bash
yarn lint
# yarn lint should show
/home/me/Projects/FreshReactNative/App.js
  29:7   warning  'test' is assigned a value but never used  no-unused-vars
  29:14  error    'bool' is not defined                      no-undef
  29:19  warning  Expected '===' and instead saw '=='        eqeqeq

✖ 3 problems (1 error, 2 warnings)
```

@fabriziobertoglio1987 Could you verify that you can reproduce these steps and they are working for you?

fabriziobertoglio1987 commented:
```
git clone git@github.com:facebook/react-native.git
cd react-native
yarn install
yarn lint
```

what I'm doing wrong? thanks

likern commented:
@fabriziobertoglio1987 Thanks for pointing out to a problem, I updated PR to make it working with this repository itself.

`yarn lint` now working identically as previously on `react-native` repository on `master` branch.

1.  This is a `yarn lint` command output for react-native `master` branch https://gist.github.com/likern/c3523de041ff7550aab555fcf2e45570

2. This is a `yarn lint` command output for react-native with changes in this pull request https://gist.github.com/likern/13b627c28a96a3701a481942c4ee806d

They are almost identical, except this warning
```bash
/home/me/test-react-native/packages/rn-tester/js/examples/NativeAnimation/NativeAnimationsExample.js
   66:19  warning  Script URL is a form of eval  no-script-url
  344:19  warning  Script URL is a form of eval  no-script-url
```

These warnings are false-positives, it's this line of code
```jsx
<Text>JavaScript:</Text>
```

Also I've run all tests using `yarn test` and tried to identify all failed tests which are possibly related to code in this PR.
```bash
Test Suites: 2 failed, 1 skipped, 134 passed, 136 of 137 total
Tests:       31 failed, 3 skipped, 1958 passed, 1992 total
Snapshots:   643 passed, 643 total
Time:        13.946 s
Ran all test suites.
```

I've fixed them (changed `babel-eslint` to `@babel/eslint-parser` and they got working). I believe other failed tests not related to this PR.

fabriziobertoglio1987 commented:
sorry. I will try to test again end of this week but I can not review your pr because I don't have expertise in this field. Thanks a lot for your contribution. :pray: :peace_symbol:

ls-andrew-goodale commented on package.json:
```diff
@@ -88,6 +88,8 @@
   },
   "dependencies": {
+    "@babel/eslint-parser": "^7.13.10",
+    "@babel/eslint-plugin": "^7.13.10",
```
I think these should be `devDependencies` since they're for lint that runs when you're working on react-native itself? The dependencies for users of react-native come down as part of the `@react-native-community/eslint-config` install.

sota000 reviewed (changes requested):
Hi @likern, can you look into solving the merge conflict? Then I'll test our internal workflow so that we can land this PR.

likern commented:
Unfortunately I don't have time to resolve conflicts, busy on other projects. If someone wants to pick up this, you are welcome.

km-tr commented:
@likern 
If there is no one else, I am interested.

likern commented:
@km-tr I did just update, react-native is using eslint-config-prettier so I just updated version. Otherwise more intrusive changes might break RN devs workflow.

likern commented:
> @likern If you like, I've resolved the conflicts in this branch. I tried to submit a PR to breeffy, but hesitated because of the large difference [km-tr:update/eslint-config-react-native-community-conflict](https://github.com/km-tr/react-native/tree/update/eslint-config-react-native-community-conflict)

You can directly commit to this PR, can you?

km-tr commented:
@likern 
I tried, but I don't have permission to commit to breeffy:update/eslint-config-react-native-community.

cortinico commented:
Closing as we're now on `"@babel/eslint-parser": "^7.19.0"` and this is not relevant anymore. Thanks for the contribution anyway @breeffy
